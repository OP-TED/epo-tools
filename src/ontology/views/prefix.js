const hasPrefix = (name) => name ? name.split(':').length === 2 : false

const getPrefix = (name) => hasPrefix(name) ? name.split(':')[0] : undefined

const stripPrefix = (str) => str.split(':').slice(1).join(':') || str

function getAllPrefixes ({ nodes, edges }) {
  return noDuplicates(nodes.map(x => x.name).filter(hasPrefix).map(getPrefix))
}

const noDuplicates = (arr) => [...new Set(arr)]

// modified from: https://www.30secondsofcode.org/js/s/string-case-conversion/
const toSpaced = str => str && str.match(
  /[A-Z]{2,}(?=[A-Z][a-z]+[0-9]*|\b)|[A-Z]?[a-z]+[0-9]*|[A-Z]|[0-9]+/g).
  // map(x => str.toUpperCase() === str ? x : x.toLowerCase()).
  map(x => x.toLowerCase()).
  join(' ')

export { hasPrefix, getPrefix, stripPrefix, getAllPrefixes, toSpaced }
