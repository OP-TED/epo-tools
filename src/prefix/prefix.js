import { aliases, ns } from '../namespaces.js'

const hasPrefix = (name) => name ? name.split(':').length === 2 : false

const getPrefix = (name) => hasPrefix(name) ? name.split(':')[0] : undefined

const stripPrefix = (str) => str.split(':').slice(1).join(':') || str

const noDuplicates = (arr) => [...new Set(arr)]

function getAllPrefixes ({ nodes, edges }) {
  return noDuplicates(nodes.map(x => x.name).filter(hasPrefix).map(getPrefix))
}

// modified from: https://www.30secondsofcode.org/js/s/string-case-conversion/
const toSpaced = str => str && str.match(
  /[A-Z]{2,}(?=[A-Z][a-z]+[0-9]*|\b)|[A-Z]?[a-z]+[0-9]*|[A-Z]|[0-9]+/g).
  // map(x => str.toUpperCase() === str ? x : x.toLowerCase()).
  map(x => x.toLowerCase()).
  join(' ')

const allPrefixes = { ...ns, ...aliases }

function toTerm (value) {
  if (!hasPrefix(value)) {
    throw Error(`${value} has no prefix`)
  }
  const prefix = getPrefix(value)
  if (!allPrefixes[prefix]) {
    throw Error(`${prefix} not known (${value})`)
  }
  return allPrefixes[prefix](stripPrefix(value))
}

export { hasPrefix, getPrefix, stripPrefix, getAllPrefixes, toSpaced, toTerm }
