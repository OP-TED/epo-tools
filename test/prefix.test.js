import { expect } from 'expect'
import toMatchSnapshot from 'expect-mocha-snapshot'
import { describe, it } from 'mocha'
import { getAllPrefixes } from '../src/prefix/prefix.js'

import eaJson from '../public/models/ePO_CM_v4.1.1.json' with { type: 'json' }

expect.extend({ toMatchSnapshot })

describe('prefix', () => {

  it(`getAllPrefixes`, function () {
    const allPrefixes = getAllPrefixes(eaJson)
    expect(allPrefixes).toMatchSnapshot(this)
  })
})

