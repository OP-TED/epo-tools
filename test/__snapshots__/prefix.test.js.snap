// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prefix getAllPrefixes 1`] = `
Array [
  "epo",
  "epo-sub",
  "at-voc",
  "epo-ord",
  "epo-ful",
  "at-voc-new",
  "locn",
  "foaf",
  "cpov",
  "epo-not",
  "org",
  "dct",
  "cccev",
  "epo-con",
  "person",
  "epo-cat",
  "time",
  "adms",
  "epo-acc",
  "eli",
  "cv",
]
`;
